src/session.ts

```diff
@@ -163,28 +168,30 @@ class NotebookSession {
       if (success.xhr.status !== 204) {
         throw Error('Invalid response');
       }
-      validateSessionId(success.data);
+      validate.validateSessionId(success.data);
     }, (rejected: utils.IAjaxError) => {
-        if (rejected.xhr.status === 410) {
-          throw Error('The kernel was deleted but the session was not');
-        }
-        throw Error(rejected.statusText);
+      if (rejected.xhr.status === 410) {
+        throw Error('The kernel was deleted but the session was not');
+      }
+      throw Error(rejected.statusText);
     });
   }
 
   /**
    * Restart the session by deleting it and then starting it fresh.
    */
-  restart(options?: ISessionOptions): Promise<void> {
-    return this.delete().then(() => this.start()).catch(
-        () => this.start()).then(() => {
-      if (options && options.notebookPath) {
-        this._notebookPath = options.notebookPath;
-      }
-      if (options && options.kernelName) {
-        this._kernel.name = options.kernelName;
-      }
-    })
+  restart(options?: ISessionOptions): Promise<ISessionId> {
+    return this.delete().then(() => {
+      return this.start().then((id: ISessionId) => {
+        if (options && options.notebookPath) {
+          this._notebookPath = options.notebookPath;
+        }
+        if (options && options.kernelName) {
+          this._kernel.name = options.kernelName;
+        }
+        return id;
+      });
+    });
```

Comment:
The options handling definitely doesn't need to be preserved as-is. It was a convenient, lazy way to hand around common options like base_url, etc. without thinking about where they are actually needed. I think @jasongrout may have used it to give extra options to alternate implementations like the sage single-cell server, but I think we can start over with only the parameters we actually use, and think differently about how alternate implementations are managed.

I do think that functions taking more than one or two args should take keyword-args and have default values. In ES5 conventions, this means having a positional arg that's a dict, typically called `options`. I'm not sure how typescript/es6 tools compile default parameters and kwargs to ES5. I think they probably just force positional-only. That seems crummy, but I don't think we will be consuming these APIs with ES5, so maybe it's okay.